### ISSUE TITLE
main

### ISSUE THREAD

**Issue by ZhigangSu** [ZhigangSu/Zhigang-Su#12]
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader, Subset
import random

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

train_dir = r"C:\Users\Ray\Documents\PlantDiseaseProject\dataset\train"
test_dir = r"C:\Users\Ray\Documents\PlantDiseaseProject\dataset\test"

transform = transforms.Compose([
    transforms.Resize((128, 128)),
    transforms.ToTensor()
])

full_train_dataset = datasets.ImageFolder(train_dir, transform=transform)
test_dataset = datasets.ImageFolder(test_dir, transform=transform)

def get_subset_by_class(dataset, max_per_class=500):
    class_indices = {i: [] for i in range(len(dataset.classes))}
    for idx, (img, label) in enumerate(dataset):
        if len(class_indices[label]) < max_per_class:
            class_indices[label].append(idx)
    selected_indices = [idx for indices in class_indices.values() for idx in indices]
    return Subset(dataset, selected_indices)

train_dataset = get_subset_by_class(full_train_dataset, max_per_class=500)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32)

class LeafDiseaseCNN(nn.Module):
    def __init__(self, num_classes):
        super(LeafDiseaseCNN, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 16, 3, padding=1), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(16, 32, 3, padding=1), nn.ReLU(), nn.MaxPool2d(2)
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(32 * 32 * 32, 128), nn.ReLU(),
            nn.Linear(128, num_classes)
        )

    def forward(self, x):
        x = self.features(x)
        return self.classifier(x)

model = LeafDiseaseCNN(num_classes=2).to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

print(" Quick Training started...")
for epoch in range(3):
    model.train()
    running_loss = 0.0
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    print(f"Epoch [{epoch+1}/3], Loss: {running_loss:.4f}")

model.eval()
correct = 0
total = 0
with torch.no_grad():
    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

accuracy = 100 * correct / total
print(f" Quick Test Accuracy: {accuracy:.2f}%")

torch.save(model.state_dict(), "leaf_disease_model.pth")
print(" Model saved as 'leaf_disease_model.pth'")
